File: regiondeployer/Deployer.cs
using System;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.Azure.Management.ResourceManager.Fluent;
using Microsoft.Azure.Management.ResourceManager.Fluent.Core;
using Microsoft.Azure.Management.Fluent;
using Microsoft.Azure.Management.ResourceManager.Fluent.Models;
using Microsoft.Azure.Management.AppService.Fluent;
using Microsoft.Rest.Azure;
using Microsoft.Azure.Management.Storage.Fluent;
using Newtonsoft.Json;

namespace regiondeployer
{
    class RegionDeployer
    {
        private static int concurrentDeploymentsLimit = AzureRegions.Load(azureRegionsJsonPath).Length;

        private static HttpResponseMessage PostDeploymentRequest(string appName, string zipFilePath, string username, string password, Action<string> logMessage)
        {
            using (var http = new HttpClient())
            {
                logMessage("making post request to apps kudu interface...");
                var filesBytes = File.ReadAllBytes(zipFilePath);
                var httpContent = new ByteArrayContent(filesBytes);
                httpContent.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
                var basicAuthPair = HttpUtilities.CreateBasicAuthHeader(username, password);

                var address = $"https://{appName}.scm.azurewebsites.net/api/zipdeploy";

                var requestMessage = HttpUtilities.CreateRequestMessage(address, HttpMethod.Post, httpContent, basicAuthPair);

                var output = http.SendAsync(requestMessage).Result;

                requestMessage.Dispose();

                logMessage("response received.");
                return output;
            }
        }

        public static (IFunctionApp, bool) DeployOne(string appName, string regionName, string zipFilePath, string resourceGroupName, IStorageAccount storageAccount, Action<string> logMessage)
        {
            var farmer = CreateFarmer(appName, regionName, resourceGroupName, storageAccount, logMessage);
            farmer.Stop(); // Prevent file copy locking issues.

            logMessage($"deploying regional farmer...{appName}{regionName}");

            var publishingProfile = farmer.GetPublishingProfile();
            var username = publishingProfile.GitUsername;
            var password = publishingProfile.GitPassword;

            var response = PostDeploymentRequest(farmer.Name, zipFilePath, username, password, logMessage);

            if (!response.IsSuccessStatusCode)
            {
                logMessage($"deployment failed for farmer {farmer.Name} {response.StatusCode} {response.ReasonPhrase}");
            }
            else
            {
                logMessage($"deployed to regional farmer. {regionName}");
            }
            farmer.Start();
            response.Dispose();
            return (farmer, response.IsSuccessStatusCode);
        }

        public static bool DeployArray(Arguments inputs, string zipFilePath, bool failOnError, string resourceGroupName, Action<string> logMessage, IStorageAccount storageAccount)
        {
            logMessage($"deploying the array of regional farmers...{inputs.NumberOfFarmers} long.");

            var regions = AzureRegions.Load(azureRegionsJsonPath);

            int deploymentIndex = 0;
            int successfulDeployments = 0;

            while (deploymentIndex < inputs.NumberOfFarmers)
            {
                int slotSize = Math.Min(inputs.NumberOfFarmers - deploymentIndex, concurrentDeploymentsLimit);
                int slotEndIndex = deploymentIndex + slotSize;

                logMessage($"deploying farmers {deploymentIndex} up to {slotEndIndex}");

                Parallel.For(deploymentIndex, slotEndIndex, i =>
                {
                    try
                    {
                        logMessage($"deploying farmer {i}");

                        var currentRegion = i < regions.Length ? regions[i].Name : defaultRegion;
                        var randomFarmerName = GetRandomFarmerName(logMessage);

                        var (createdResource, deploymentStatus) = DeployOne(randomFarmerName, currentRegion, zipFilePath, resourceGroupName, storageAccount, logMessage);

                        if (deploymentStatus)
                        {
                            var masterKey = createdResource.GetMasterKey();
                            DatabaseSeeder.SeedNewRegionalFarmer(createdResource, masterKey, logMessage);
                            logMessage($"region farmer {i} deployed. {currentRegion}");
                            successfulDeployments++;
                        }
                    }
                    catch (Exception error)
                    {
                        if (failOnError)
                        {
                            PrintCloudError(error, logMessage);
                            throw;
                        }
                        else
                        {
                            PrintCloudError(error, logMessage);
                        }
                    }
                });

                deploymentIndex = slotEndIndex;
            }

            logMessage($"array deployed. there were {successfulDeployments}/{inputs.NumberOfFarmers} successful deployments.");
            return true;
        }
    }
}

File: regiondeployer/deployer.fs
module regiondeployer.deployer

open Microsoft.Azure.Management.ResourceManager.Fluent
open Microsoft.Azure.Management.ResourceManager.Fluent.Core
open Microsoft.Azure.Management.Fluent
open System
open System.Linq
open Microsoft.Azure.Management.ResourceManager.Fluent.Models
open Microsoft.Azure.Management.AppService.Fluent
open System.Configuration
open regiondeployer.constants.independentconstants
open regiondeployer.cloudactions.azurefunction.farmer
open regiondeployer.cloudactions.storage
open System.Net.Http
open System.IO.Compression
open System.IO
open System.Text
open regiondeployer.azureauthenticator
open Microsoft.Rest.Azure
open regiondeployer.cloudactions.resourcegroup
open regiondeployer.types
open regiondeployer.databaseseeder
open System.Threading.Tasks
open System.Threading
open Microsoft.Azure.Management.Storage.Fluent
open regiondeployer
open regiondeployer.randomnamegenerator
open regiondeployer.clouderror
open regiondeployer.constants.pathconstants
open Newtonsoft.Json
open serversidearchitecture.common
open System.Collections.Generic
open System.Net.Http.Headers

let private concurrentdeploymentslimit = azureregions.Load(azureregionsjsonpath).Length

let private postdeploymentrequest (appname:string, zipfilepath:string, username:string, password:string, logmessage:string -> unit) : HttpResponseMessage =   
        use http = new HttpClient()
        logmessage "making post request to apps kudu interface..."
        let filesbytes = File.ReadAllBytes(zipfilepath)
        let httpcontent = new ByteArrayContent(filesbytes)
        httpcontent.Headers.ContentType <- new MediaTypeHeaderValue("application/octet-stream")
        let basicauthpair : KeyValuePair<string, string> = http.createbasicauthheader(username, password)  

        let address = 
            String.concat 
                String.Empty [@"https://"; appname; ".scm.azurewebsites.net/api/zipdeploy"]

        let requestmessage:HttpRequestMessage = 
            httputilities.createrequestmessage(address, HttpMethod.Post, httpcontent, basicauthpair)

        let output = 
            (globalhttpport.get
                .client
                .SendAsync(requestmessage)
                .Result)

        requestmessage.Dispose()

        logmessage "response received."
        output

///returns the creates azure resource and the status of deployment
let public deployone (appname:string, regionname:string, zipfilepath:string, resourcegroupname:string, storageaccount:IStorageAccount, logmessage:string -> unit) : (IFunctionApp * bool) =    
    let farmer = createfarmer(appname, regionname, resourcegroupname, storageaccount, logmessage)
    farmer.Stop() //should prevent file copy locking issues. preventative programming.

    String.concat valueseparator ["deploying regional farmer..."; appname; regionname]
    |> logmessage

    let publishingprofile = farmer.GetPublishingProfile()
    let username = publishingprofile.GitUsername
    let password = publishingprofile.GitPassword

    let response = postdeploymentrequest(farmer.Name, zipfilepath, username, password, logmessage)

    if response.IsSuccessStatusCode = false then
        String.concat valueseparator [
            "deployment failed for farmer"; 
            farmer.Name; 
            response.StatusCode.ToString(); 
            response.ReasonPhrase;            
        ]
        |> logmessage

    else
        logmessage (
            String.concat valueseparator [ "deployed to regional farmer."; regionname]
        )
    farmer.Start()
    response.Dispose()
    (farmer, response.IsSuccessStatusCode)

let public deployarray (inputs:Arguments, zipfilepath:string, failonerror:bool, resourcegroupname:string, logmessage:string -> unit, storageaccount:IStorageAccount) : bool =
    String.concat valueseparator [ 
        "deploying the array of regional farmers..."; 
        inputs.NumberOfFarmers.ToString(); 
        "long."]
    |> logmessage
    
    let regions = azureregions.Load(azureregionsjsonpath)
    
    let mutable deploymentindex = 0
    let mutable successfuldeployments = 0

    let deployfarmersinspecificrange (startindexinclusive:int) (endindexexclusive:int) (inputs:Arguments) : ParallelLoopResult =
        Parallel.For(startindexinclusive, endindexexclusive, fun i state ->    
            try
                logmessage (
                    String.concat 
                        valueseparator 
                        ["deploying farmer"; i.ToString()]
                )

                let currentregion =
                    if i < regions.Length then
                        regions.[i].Name //somehow azure converts input Name into output Display Name. I dont think its worth risking testing input display name.

                    else
                        defaultregion

                let randomfarmername = 
                    (getrandomfarmername logmessage)

                deployone(            
                    randomfarmername,
                    currentregion,
                    zipfilepath,         
                    resourcegroupname,
                    storageaccount,
                    logmessage
                )
                |> fun resultwip -> match resultwip with
                | (createdresource, deploymentstatus) ->    
                    let masterkey = createdresource.GetMasterKey()
                    databaseseeder.seednewregionalfarmer (createdresource, masterkey, logmessage)
                    String.concat valueseparator ["regionfarmer"; i.ToString(); "deployed."; currentregion]
                    |> logmessage 

                    if deploymentstatus = true then
                        successfuldeployments <- successfuldeployments + 1                

            with 
            | error1 -> //parallel.for smothers errors occuring in its worker threads
                if failonerror then
                    printclouderror(error1, logmessage)
                    raiseclouderror error1                    
        
                else
                    printclouderror(error1, logmessage)
                    
                ()
        )    
    
    try
        while deploymentindex < inputs.NumberOfFarmers do
            let slotsize = 
                let workleft = inputs.NumberOfFarmers - deploymentindex

                if workleft < concurrentdeploymentslimit then 
                    workleft

                else
                    concurrentdeploymentslimit

            let slotendindex = deploymentindex + slotsize

            String.concat valueseparator ["deploying farmers"; deploymentindex.ToString(); "up to"; slotendindex.ToString()]
            |> logmessage

            deployfarmersinspecificrange deploymentindex slotendindex inputs
            |> fun result ->    
                deploymentindex <- slotendindex //the end index becomes the new index since it was excluded in the previous rangedeploy

        String.concat valueseparator [
            "array deployed. there were"; 
            successfuldeployments.ToString(); 
            "/";
            inputs.NumberOfFarmers.ToString();
            "successful deployments."
        ]
        |> logmessage
        true

    with
    | error1 -> 
        if failonerror then
            printclouderror(error1, logmessage)
            raiseclouderror error1
            false
        
        else
            printclouderror(error1, logmessage)
            false


